File: src/day04.rs
use crate::lib::to_filename;

use std::fs;

use counter::Counter;

type Num = u32;

fn get_data() -> impl Iterator<Item = Num> {
    fs::read_to_string(to_filename(4))
        .expect("Could not read file")
        .lines()
        .map(|s| read_row(s))
        .next()
        .unwrap()
}

fn read_row(row: &str) -> impl Iterator<Item = Num> {
    let vals: Vec<_> = row.split('-').map(|s| s.parse::<Num>().unwrap()).collect();
    (vals[0]..=vals[1])
}

fn valid_old(n: &Num) -> bool {
    let digits = n.to_string();
    let sorted_ = digits.chars().is_sorted();

    let dupes = digits
        .chars()
        .collect::<Counter<_>>()
        .values()
        .max()
        .unwrap()
        >= &2;

    sorted_ && dupes
}

fn valid(n: &Num) -> bool {
    let digits = n.to_string();
    let sorted_ = digits.chars().is_sorted();

    let double = digits
        .chars()
        .collect::<Counter<_>>()
        .values()
        .collect::<Vec<_>>()
        .contains(&&2);

    sorted_ && double
}

pub fn part1() -> usize {
    let posses: Vec<_> = get_data().filter(valid_old).collect();
    posses.len()
}

pub fn part2() -> usize {
    let posses: Vec<_> = get_data().filter(valid).collect();
    posses.len()
}

File: src/day04.py
from itertools import groupby

from collections.abc import Iterable

from generic import ilen, flist, get_file

password = list[int]


def c0(p: password) -> bool:
    return max(map(ilen, (v for k, v in groupby(p)))) >= 2


def c1(p: password) -> bool:
    return p == sorted(p)

def c2(p: password) -> bool:
    return 2 in (map(ilen, (v for k, v in groupby(p))))

def posses(s: str) -> Iterable[password]:
    base = range(*map(int, s.split("-")))
    return (list(map(int, str(i))) for i in base)


def valid1(p: password):
    return all(flist([c0, c1], p))

def valid2(p: password):
    return all(flist([c2, c1], p))


def part1(s: str) -> int:
    return ilen(filter(valid1, posses(s)))

def part2(s: str) -> int:
    return ilen(filter(valid2, posses(s)))

def main():
    s = get_file(4)
    print(f"{part1(s)=}")
    print(f"{part2(s)=}")


if __name__ == "__main__":
    main()
